File: src/general/json.js
function completeTruncatedJSON(inputStr, multiple_jsons=false) {
    if (!inputStr) return null;
    inputStr = inputStr.trim();
    if(multiple_jsons) inputStr = "[" + inputStr;

    let stack = [];
    let insideString = false;

    for (let i=0; i<inputStr.length; i++) {
        let char = inputStr[i];
        if (char === '"' && ((i > 0 && inputStr[i - 1] !== '\\') || (i >1 && inputStr[i - 2] === '\\'))) {
            insideString = !insideString;
        }

        if (!insideString) {
            if (char === '{' || char === '[') {
                stack.push(char);
            } else if (char === ':') {
                stack.pop();
                stack.push(":");
            } else if (char === ',' && stack[stack.length-1] === ':') {
                stack.pop();
                stack.push("{");
            } else if (char === '}' && stack[stack.length - 1] === ':') {
                stack.pop();
                if(stack.length == 1 && multiple_jsons){
                    inputStr = inputStr.slice(0, i+1) + "," + inputStr.slice(i+1);
                }
            } else if (char === ']' && stack[stack.length - 1] === '[') {
                stack.pop();
                if(stack.length == 1 && multiple_jsons){
                    inputStr = inputStr.slice(0, i+1) + "," + inputStr.slice(i+1);
                }
            }
        }
    }

    while (true) {
        try {
            let json = JSON.parse(inputStr);
            removeIncomplete(json);
            return json;
        } catch (e) {
            if (insideString) {
                inputStr += '_unfinished_"';
                insideString = false;
                continue;
            }

            if (stack.length === 0) {
                break;
            }

            let lastOpened = stack[stack.length - 1];

            if (lastOpened === ':') {
                if (inputStr.endsWith(':')) {
                    inputStr += '"_unfinished_"}';
                } else if (inputStr.endsWith(',')) {
                    inputStr = inputStr.slice(0, -1) + '}';
                } else {
                    inputStr += '}';
                }
                stack.pop();
            } else if (lastOpened === '{') {
                if (inputStr.endsWith('{')) {
                    inputStr += '"_unfinished_": "_unfinished_"}';
                } else if (inputStr.endsWith(',')) {
                    inputStr = inputStr.slice(0, -1) + '}';
                } else {
                    inputStr += ': "_unfinished_"}';
                }
                stack.pop();
            } else if (lastOpened === '[') {
                if (inputStr.endsWith('[')) {
                    inputStr += '"_unfinished_"]';
                } else if (inputStr.endsWith(',')) {
                    inputStr = inputStr.slice(0, -1) + ']';
                } else {
                    inputStr += ']';
                }
                stack.pop();
            }
        }
    }

    return null;
}

function removeIncomplete(data) {
    if (Array.isArray(data)) {
        let i = data.length - 1;
        if (isIncomplete(data[i])) {
            data.splice(i, 1);
            i--;  // adjust index after splicing
        }  
    } else if (typeof data === 'object' && data !== null) {
        let key = Object.keys(data)[Object.keys(data).length - 1]
        if (isIncomplete(key) || isIncomplete(data[key])) {
            delete data[key];
        } else {
            removeIncomplete(data[key])
        }
    }
}

function isIncomplete(value) {
    if (typeof value === 'string') {
        // Check for trailing spaces or if it's a placeholder value
        return value.endsWith('_unfinished_');
    } else if (!Array.isArray(value) && typeof value === 'object') {
        if (value === null) return false;

        for (let key in value) {
            if (isIncomplete(key) || isIncomplete(value[key])){
                return true;
            } 
        }
    }
    return false;
}


export { completeTruncatedJSON };

File: src/utils/json.py
import json

def complete_truncated_json(input_str, multiple_jsons=False):
    if not input_str:
        return None
    input_str = input_str.strip()
    
    if multiple_jsons:
        input_str = "[" + input_str

    stack = []
    inside_string = False

    i = 0
    while i < len(input_str):
        char = input_str[i]
        
        if char == '"' and (i == 0 or input_str[i - 1] != '\\'):
            inside_string = not inside_string
        
        if not inside_string:
            if char == '{' or char == '[':
                stack.append(char)
            elif char == ':':
                stack.pop()
                stack.append(":")
            elif char == ',' and stack[-1] == ':':
                stack.pop()
                stack.append("{")
            elif char == '}' and stack[-1] == ':':
                stack.pop()
                if len(stack) == 1 and multiple_jsons:
                    input_str = input_str[:i+1] + "," + input_str[i+1:]
            elif char == ']' and stack[-1] == '[':
                stack.pop()
                if len(stack) == 1 and multiple_jsons:
                    input_str = input_str[:i+1] + "," + input_str[i+1:]
        
        i += 1

    while True:
        try:
            parsed_json = json.loads(input_str)
            remove_incomplete(parsed_json)
            return parsed_json
        except json.JSONDecodeError:
            if inside_string:
                input_str += '_unfinished_"'
                inside_string = False
                continue
            
            if not stack:
                break
            
            last_opened = stack[-1]
            
            if last_opened == ':':
                if input_str.endswith(':'):
                    input_str += '"_unfinished_"}'
                elif input_str.endswith(','):
                    input_str = input_str[:-1] + '}'
                else:
                    input_str += '}'
                stack.pop()
            elif last_opened == '{':
                if input_str.endswith('{'):
                    input_str += '"_unfinished_": "_unfinished_"}'
                elif input_str.endswith(','):
                    input_str = input_str[:-1] + '}'
                else:
                    input_str += ': "_unfinished_"}'
                stack.pop()
            elif last_opened == '[':
                if input_str.endswith('['):
                    input_str += '"_unfinished_"]'
                elif input_str.endswith(','):
                    input_str = input_str[:-1] + ']'
                else:
                    input_str += ']'
                stack.pop()

    return None


def remove_incomplete(data):
    if isinstance(data, list):
        if len(data) > 0 and is_incomplete(data[-1]):
            data.pop()
    elif isinstance(data, dict):
        last_key = list(data.keys())[-1]
        if is_incomplete(last_key) or is_incomplete(data[last_key]):
            del data[last_key]
        else:
            remove_incomplete(data[last_key])


def is_incomplete(value):
    if isinstance(value, str):
        return value.endswith('_unfinished_')
    elif isinstance(value, dict):
        for key, val in value.items():
            if is_incomplete(key) or is_incomplete(val):
                return True
    return False
